{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "from numba import njit\n",
    "import numpy as np\n",
    "%run tools.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "@njit()\n",
    "def scanRight(geno1, geno2, upperBreakpoint, max): \n",
    "    # Reset position for next loop\n",
    "    gn1 = geno1[upperBreakpoint]\n",
    "    gn2 = geno2[upperBreakpoint]\n",
    "\n",
    "    # Scan right along genome, as long as two inds are not both homozygous but different\n",
    "    while not (gn1[0] == gn1[1]) & (gn2[0] == gn2[1]) & ((gn1 != gn2).all()) & (-1 not in gn1 and -1 not in gn2):\n",
    "\n",
    "        upperBreakpoint += 1\n",
    "        if upperBreakpoint == max-1: # limit the upper breakpoint at end of the contig\n",
    "            return(upperBreakpoint)\n",
    "\n",
    "        gn1 = geno1[upperBreakpoint]\n",
    "        gn2 = geno2[upperBreakpoint]\n",
    "        \n",
    "    return(upperBreakpoint)\n",
    "\n",
    "@njit()\n",
    "def scanLeft(geno1, geno2, lowerBreakpoint): \n",
    "\n",
    "    # subset genotypes to this position, because we need somethign to start the while loop?\n",
    "    gn1 = geno1[lowerBreakpoint]\n",
    "    gn2 = geno2[lowerBreakpoint]\n",
    "\n",
    "    # Scan left along genome\n",
    "    while not (gn1[0] == gn1[1]) & (gn2[0] == gn2[1]) & ((gn1 != gn2).all()) & (-1 not in gn1 and -1 not in gn2):\n",
    "\n",
    "        lowerBreakpoint -= 1\n",
    "        if lowerBreakpoint == 0: # limit lower breakpoint at zero, start of contig\n",
    "            return(lowerBreakpoint)\n",
    "\n",
    "        gn1 = geno1[lowerBreakpoint]\n",
    "        gn2 = geno2[lowerBreakpoint]\n",
    "        \n",
    "    return(lowerBreakpoint)\n",
    "\n",
    "\n",
    "@njit()\n",
    "def f2scans(dblton_arr, snps, pos):\n",
    "\n",
    "    starts = []# np.empty((len(dblton_arr)),dtype='uint8')\n",
    "    ends = []# np.empty((len(dblton_arr)),dtype='uint8')\n",
    "    dbltonpos = []\n",
    "\n",
    "    for idx in range(0, len(dblton_arr)):\n",
    "\n",
    "        geno1 = snps[:, dblton_arr[idx][0]]\n",
    "        geno2 = snps[:, dblton_arr[idx][1]]\n",
    "        # get boolean of dblton idx\n",
    "        #dblton_idx = bisect.bisect_left(pos, dblton_arr[idx][2])\n",
    "        dblton_idx = np.searchsorted(pos, dblton_arr[idx][2])\n",
    "\n",
    "        upperBreakpoint = scanRight(geno1, geno2, dblton_idx, len(pos))\n",
    "        # Scan left along genome\n",
    "        lowerBreakpoint = scanLeft(geno1, geno2, dblton_idx)\n",
    "\n",
    "        starts.append(pos[lowerBreakpoint])\n",
    "        ends.append(pos[upperBreakpoint])\n",
    "        dbltonpos.append(dblton_arr[idx][2])\n",
    "    \n",
    "    return(np.array(starts), np.array(ends), np.array(dbltonpos))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"../../config/metadata.tsv\", sep=\"\\t\")\n",
    "dblton = pd.read_csv(\"../../results/f2variantPairs.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "contigs = ['2L', '2R', '3L', '3R', 'X']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "    # Load Arrays\n",
    "contig = 'X'\n",
    "\n",
    "snps, pos = loadZarrArrays(genotypePath=f\"../../resources/snp_genotypes/all/1244-VO-GH-YAWSON-VMF00149/{contig}/calldata/GT/\", \n",
    "                                        positionsPath=f\"../../resources/snp_genotypes/all/sites/{contig}/variants/POS/\",\n",
    "                                        siteFilterPath=f\"../../resources/site_filters/dt_20200416/gamb_colu/{contig}/variants/filter_pass/\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "singletons = snps.count_alleles().is_singleton()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "singletons = singletons.compute()\n",
    "pos_sin = pos[singletons]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps.compress(singletons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs1D\"><span>&lt;SortedIndex shape=(673176,) dtype=int32&gt;</span><table><thead><tr><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">673171</th><th style=\"text-align: center\">673172</th><th style=\"text-align: center\">673173</th><th style=\"text-align: center\">673174</th><th style=\"text-align: center\">673175</th></tr></thead><tbody><tr><td style=\"text-align: center\">30</td><td style=\"text-align: center\">53</td><td style=\"text-align: center\">58</td><td style=\"text-align: center\">128</td><td style=\"text-align: center\">165</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">24390291</td><td style=\"text-align: center\">24390293</td><td style=\"text-align: center\">24390312</td><td style=\"text-align: center\">24390481</td><td style=\"text-align: center\">24392596</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<SortedIndex shape=(673176,) dtype=int32>\n",
       "[30, 53, 58, 128, 165, ..., 24390291, 24390293, 24390312, 24390481, 24392596]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_sin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2L (439698, 3)\n",
      "Scanning 2L doubletons\n",
      "Done...\n",
      "2R (576303, 3)\n",
      "Scanning 2R doubletons\n",
      "Done...\n",
      "3L (440880, 3)\n",
      "Scanning 3L doubletons\n",
      "Done...\n",
      "3R (615762, 3)\n",
      "Scanning 3R doubletons\n",
      "Done...\n",
      "X (209303, 3)\n",
      "Scanning X doubletons\n",
      "Done...\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "for contig in contigs:\n",
    "    snps = {}\n",
    "    pos = {}\n",
    "\n",
    "    # Load Arrays\n",
    "    snps, pos = loadZarrArrays(genotypePath=f\"../../resources/snp_genotypes/all/1244-VO-GH-YAWSON-VMF00149/{contig}/calldata/GT/\", \n",
    "                                            positionsPath=f\"../../resources/snp_genotypes/all/sites/{contig}/variants/POS/\",\n",
    "                                            siteFilterPath=f\"../../resources/site_filters/dt_20200416/gamb_colu/{contig}/variants/filter_pass/\")\n",
    "    \n",
    "\n",
    "    ac = snps.count_alleles()\n",
    "    seg = ac.is_segregating()\n",
    "    snps = snps.compress(seg, axis=0).compute().values\n",
    "    pos = pos[seg]\n",
    "    ### Load doubletons\n",
    "\n",
    "    dblton_arr = dblton.query(\"contig == @contig\")[['idx1', 'idx2', 'pos']].to_numpy()\n",
    "    log(contig, dblton_arr.shape)\n",
    "\n",
    "    log(f\"Scanning {contig} doubletons\")\n",
    "    starts, ends, dbltonpos = f2scans(dblton_arr, snps, pos)\n",
    "    log(\"Done...\")\n",
    "    f2_df = pd.DataFrame({'start':starts, 'end':ends, 'dblton_pos': dbltonpos})\n",
    "    f2_df.to_csv(f\"../../results/f2variants/f2HapLengths.{contig}.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing the functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 9\n"
     ]
    }
   ],
   "source": [
    "qpos = np.array([0,10,20,30,40,50,60,70,80,90])\n",
    "dbpos = 50\n",
    "geno1 = np.array([[0,0], [-1,-1], [0,1], [0,0], [0,0], [0,1], [0,0], [0,0], [-1,-1], [0, 0]])\n",
    "geno2 = np.array([[1,1], [-1,-1], [0,0], [0,0], [0,0], [0,1], [0,0], [0,0], [0,0], [1, 1]])\n",
    "# get boolean of dblton idx\n",
    "#dblton_idx = bisect.bisect_left(pos, dblton_arr[idx][2])\n",
    "dblton_idx = np.searchsorted(qpos, dbpos)\n",
    "a = scanLeft(geno1, geno2, dblton_idx)\n",
    "b = scanRight(geno1, geno2, dblton_idx, 10)\n",
    "print(a,b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "contig = '2L'\n",
    "# Load Arrays\n",
    "snps, pos = loadZarrArrays(genotypePath=f\"../../resources/snp_genotypes/all/1244-VO-GH-YAWSON-VMF00149/{contig}/calldata/GT/\", \n",
    "                                        positionsPath=f\"../../resources/snp_genotypes/all/sites/{contig}/variants/POS/\",\n",
    "                                        siteFilterPath=f\"../../resources/site_filters/dt_20200416/gamb_colu/{contig}/variants/filter_pass/\")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac = snps.count_alleles()\n",
    "seg = ac.is_segregating()\n",
    "snps = snps.compress(seg, axis=0).compute().values\n",
    "pos = pos[seg]\n",
    "### Load doubletons\n",
    "dblton_arr = dblton.query(\"contig == @contig\")[['idx1', 'idx2', 'pos']].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "218 ms ± 6.03 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "starts, ends, dbltonpos = f2scans(dblton_arr[250_000:250_100], snps, pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ends-starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx1 = dblton_arr[1][0]\n",
    "idx2 = dblton_arr[1][1]\n",
    "posd = dblton_arr[1][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "200 ms ± 6.66 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    " %%timeit\n",
    "starts, ends, dbltonpos = f2scans(dblton_arr[250_000:250_100], snps, pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = snps.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Timings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def speedup(fullnj, half, naive):\n",
    "\n",
    "    print(f\"naive / half speedup: {naive/half}\")\n",
    "    print(f\"half/full speedup: {half/fullnj}\")\n",
    "    print(f\"overall speedup: {naive/fullnj}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "naive / half speedup: 11.755725190839694\n",
      "half/full speedup: 19.293078055964653\n",
      "overall speedup: 226.8041237113402\n"
     ]
    }
   ],
   "source": [
    "speedup(0.0679, 1.31, 15.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_filter = zarr.open_array(f\"../../resources/site_filters/dt_20200416/gamb_colu/{contig}/variants/filter_pass/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>partner_sample_id2</th>\n",
       "      <th>pos</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude2</th>\n",
       "      <th>longitude2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2L</td>\n",
       "      <td>3</td>\n",
       "      <td>198</td>\n",
       "      <td>WA-2151</td>\n",
       "      <td>WA-2225</td>\n",
       "      <td>81</td>\n",
       "      <td>6.304</td>\n",
       "      <td>-1.682</td>\n",
       "      <td>6.058</td>\n",
       "      <td>-1.391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2L</td>\n",
       "      <td>282</td>\n",
       "      <td>346</td>\n",
       "      <td>WA-2275</td>\n",
       "      <td>WA-2113</td>\n",
       "      <td>82</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "      <td>6.016</td>\n",
       "      <td>-1.813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2L</td>\n",
       "      <td>3</td>\n",
       "      <td>282</td>\n",
       "      <td>WA-2151</td>\n",
       "      <td>WA-2275</td>\n",
       "      <td>123</td>\n",
       "      <td>6.304</td>\n",
       "      <td>-1.682</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2L</td>\n",
       "      <td>324</td>\n",
       "      <td>327</td>\n",
       "      <td>WA-2293</td>\n",
       "      <td>WA-2296</td>\n",
       "      <td>142</td>\n",
       "      <td>5.971</td>\n",
       "      <td>-1.779</td>\n",
       "      <td>5.971</td>\n",
       "      <td>-1.778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2L</td>\n",
       "      <td>293</td>\n",
       "      <td>338</td>\n",
       "      <td>WA-2286</td>\n",
       "      <td>WA-2105</td>\n",
       "      <td>145</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "      <td>6.016</td>\n",
       "      <td>-1.812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715687</th>\n",
       "      <td>2L</td>\n",
       "      <td>295</td>\n",
       "      <td>372</td>\n",
       "      <td>WA-2433</td>\n",
       "      <td>WA-2467</td>\n",
       "      <td>49361656</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "      <td>6.015</td>\n",
       "      <td>-1.813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715688</th>\n",
       "      <td>2L</td>\n",
       "      <td>232</td>\n",
       "      <td>334</td>\n",
       "      <td>WA-2073</td>\n",
       "      <td>WA-2303</td>\n",
       "      <td>49361661</td>\n",
       "      <td>5.991</td>\n",
       "      <td>-1.370</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715689</th>\n",
       "      <td>2L</td>\n",
       "      <td>436</td>\n",
       "      <td>442</td>\n",
       "      <td>WA-2336</td>\n",
       "      <td>WA-2481</td>\n",
       "      <td>49361694</td>\n",
       "      <td>6.049</td>\n",
       "      <td>-1.961</td>\n",
       "      <td>6.047</td>\n",
       "      <td>-1.961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715690</th>\n",
       "      <td>2L</td>\n",
       "      <td>155</td>\n",
       "      <td>180</td>\n",
       "      <td>WA-2404</td>\n",
       "      <td>WA-2051</td>\n",
       "      <td>49361932</td>\n",
       "      <td>6.369</td>\n",
       "      <td>-1.464</td>\n",
       "      <td>6.108</td>\n",
       "      <td>-1.449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715691</th>\n",
       "      <td>2L</td>\n",
       "      <td>116</td>\n",
       "      <td>479</td>\n",
       "      <td>WA-2192</td>\n",
       "      <td>WA-2345</td>\n",
       "      <td>49361938</td>\n",
       "      <td>6.404</td>\n",
       "      <td>-1.451</td>\n",
       "      <td>6.207</td>\n",
       "      <td>-1.692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>715692 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       contig  idx1  idx2 partner_sample_id partner_sample_id2       pos  \\\n",
       "0          2L     3   198           WA-2151            WA-2225        81   \n",
       "1          2L   282   346           WA-2275            WA-2113        82   \n",
       "2          2L     3   282           WA-2151            WA-2275       123   \n",
       "3          2L   324   327           WA-2293            WA-2296       142   \n",
       "4          2L   293   338           WA-2286            WA-2105       145   \n",
       "...       ...   ...   ...               ...                ...       ...   \n",
       "715687     2L   295   372           WA-2433            WA-2467  49361656   \n",
       "715688     2L   232   334           WA-2073            WA-2303  49361661   \n",
       "715689     2L   436   442           WA-2336            WA-2481  49361694   \n",
       "715690     2L   155   180           WA-2404            WA-2051  49361932   \n",
       "715691     2L   116   479           WA-2192            WA-2345  49361938   \n",
       "\n",
       "        latitude  longitude  latitude2  longitude2  \n",
       "0          6.304     -1.682      6.058      -1.391  \n",
       "1          5.970     -1.697      6.016      -1.813  \n",
       "2          6.304     -1.682      5.970      -1.697  \n",
       "3          5.971     -1.779      5.971      -1.778  \n",
       "4          5.970     -1.697      6.016      -1.812  \n",
       "...          ...        ...        ...         ...  \n",
       "715687     5.970     -1.697      6.015      -1.813  \n",
       "715688     5.991     -1.370      5.970      -1.779  \n",
       "715689     6.049     -1.961      6.047      -1.961  \n",
       "715690     6.369     -1.464      6.108      -1.449  \n",
       "715691     6.404     -1.451      6.207      -1.692  \n",
       "\n",
       "[715692 rows x 10 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dblton"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "index 0 is out of bounds for axis 0 with size 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_5291/473626926.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'timeit'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'f2hapSize = {}\\n\\nfor row in dblton[250_000:250_100].itertuples(name=None):\\n    ## subset to individuals, we need whole chrom atm\\n    # subset genotypes to these individuals\\n    geno1 = snps[:, row[2]]\\n    geno2 = snps[:, row[3]]\\n    # get boolean of dblton idx\\n    dblton_idx = np.where(pos == row[6])[0][0]\\n\\n    upperBreakpoint = scanRight(geno1, geno2, dblton_idx, pos.shape[0])\\n\\n    # Scan left along genome\\n    lowerBreakpoint = scanLeft(geno1, geno2, dblton_idx)\\n\\n    start = pos[lowerBreakpoint]\\n    end = pos[upperBreakpoint]\\n    f2hapSize[f\"{contig}_{row[6]}\"] = end-start  # get size in bp of f2 haplotype\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2417\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2418\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2419\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2420\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2421\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/decorator.py\u001b[0m in \u001b[0;36mfun\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    230\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mkwsyntax\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m                 \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 232\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mcaller\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextras\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    233\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m   1178\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1179\u001b[0m                 \u001b[0mnumber\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m \u001b[0;34m**\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1180\u001b[0;31m                 \u001b[0mtime_number\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnumber\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1181\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtime_number\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m0.2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1182\u001b[0m                     \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, number)\u001b[0m\n\u001b[1;32m    167\u001b[0m         \u001b[0mgc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 169\u001b[0;31m             \u001b[0mtiming\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    170\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgcold\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<magic-timeit>\u001b[0m in \u001b[0;36minner\u001b[0;34m(_it, _timer)\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: index 0 is out of bounds for axis 0 with size 0"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "f2hapSize = {}\n",
    "\n",
    "for row in dblton[250_000:250_100].itertuples(name=None):\n",
    "    ## subset to individuals, we need whole chrom atm\n",
    "    # subset genotypes to these individuals\n",
    "    geno1 = snps[:, row[2]]\n",
    "    geno2 = snps[:, row[3]]\n",
    "    # get boolean of dblton idx\n",
    "    dblton_idx = np.where(pos == row[6])[0][0]\n",
    "\n",
    "    upperBreakpoint = scanRight(geno1, geno2, dblton_idx, pos.shape[0])\n",
    "\n",
    "    # Scan left along genome\n",
    "    lowerBreakpoint = scanLeft(geno1, geno2, dblton_idx)\n",
    "\n",
    "    start = pos[lowerBreakpoint]\n",
    "    end = pos[upperBreakpoint]\n",
    "    f2hapSize[f\"{contig}_{row[6]}\"] = end-start  # get size in bp of f2 haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>partner_sample_id2</th>\n",
       "      <th>pos</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude2</th>\n",
       "      <th>longitude2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2L</td>\n",
       "      <td>3</td>\n",
       "      <td>198</td>\n",
       "      <td>WA-2151</td>\n",
       "      <td>WA-2225</td>\n",
       "      <td>81</td>\n",
       "      <td>6.304</td>\n",
       "      <td>-1.682</td>\n",
       "      <td>6.058</td>\n",
       "      <td>-1.391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2L</td>\n",
       "      <td>282</td>\n",
       "      <td>346</td>\n",
       "      <td>WA-2275</td>\n",
       "      <td>WA-2113</td>\n",
       "      <td>82</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "      <td>6.016</td>\n",
       "      <td>-1.813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2L</td>\n",
       "      <td>3</td>\n",
       "      <td>282</td>\n",
       "      <td>WA-2151</td>\n",
       "      <td>WA-2275</td>\n",
       "      <td>123</td>\n",
       "      <td>6.304</td>\n",
       "      <td>-1.682</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2L</td>\n",
       "      <td>324</td>\n",
       "      <td>327</td>\n",
       "      <td>WA-2293</td>\n",
       "      <td>WA-2296</td>\n",
       "      <td>142</td>\n",
       "      <td>5.971</td>\n",
       "      <td>-1.779</td>\n",
       "      <td>5.971</td>\n",
       "      <td>-1.778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2L</td>\n",
       "      <td>293</td>\n",
       "      <td>338</td>\n",
       "      <td>WA-2286</td>\n",
       "      <td>WA-2105</td>\n",
       "      <td>145</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "      <td>6.016</td>\n",
       "      <td>-1.812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715687</th>\n",
       "      <td>2L</td>\n",
       "      <td>295</td>\n",
       "      <td>372</td>\n",
       "      <td>WA-2433</td>\n",
       "      <td>WA-2467</td>\n",
       "      <td>49361656</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.697</td>\n",
       "      <td>6.015</td>\n",
       "      <td>-1.813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715688</th>\n",
       "      <td>2L</td>\n",
       "      <td>232</td>\n",
       "      <td>334</td>\n",
       "      <td>WA-2073</td>\n",
       "      <td>WA-2303</td>\n",
       "      <td>49361661</td>\n",
       "      <td>5.991</td>\n",
       "      <td>-1.370</td>\n",
       "      <td>5.970</td>\n",
       "      <td>-1.779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715689</th>\n",
       "      <td>2L</td>\n",
       "      <td>436</td>\n",
       "      <td>442</td>\n",
       "      <td>WA-2336</td>\n",
       "      <td>WA-2481</td>\n",
       "      <td>49361694</td>\n",
       "      <td>6.049</td>\n",
       "      <td>-1.961</td>\n",
       "      <td>6.047</td>\n",
       "      <td>-1.961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715690</th>\n",
       "      <td>2L</td>\n",
       "      <td>155</td>\n",
       "      <td>180</td>\n",
       "      <td>WA-2404</td>\n",
       "      <td>WA-2051</td>\n",
       "      <td>49361932</td>\n",
       "      <td>6.369</td>\n",
       "      <td>-1.464</td>\n",
       "      <td>6.108</td>\n",
       "      <td>-1.449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715691</th>\n",
       "      <td>2L</td>\n",
       "      <td>116</td>\n",
       "      <td>479</td>\n",
       "      <td>WA-2192</td>\n",
       "      <td>WA-2345</td>\n",
       "      <td>49361938</td>\n",
       "      <td>6.404</td>\n",
       "      <td>-1.451</td>\n",
       "      <td>6.207</td>\n",
       "      <td>-1.692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>715692 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       contig  idx1  idx2 partner_sample_id partner_sample_id2       pos  \\\n",
       "0          2L     3   198           WA-2151            WA-2225        81   \n",
       "1          2L   282   346           WA-2275            WA-2113        82   \n",
       "2          2L     3   282           WA-2151            WA-2275       123   \n",
       "3          2L   324   327           WA-2293            WA-2296       142   \n",
       "4          2L   293   338           WA-2286            WA-2105       145   \n",
       "...       ...   ...   ...               ...                ...       ...   \n",
       "715687     2L   295   372           WA-2433            WA-2467  49361656   \n",
       "715688     2L   232   334           WA-2073            WA-2303  49361661   \n",
       "715689     2L   436   442           WA-2336            WA-2481  49361694   \n",
       "715690     2L   155   180           WA-2404            WA-2051  49361932   \n",
       "715691     2L   116   479           WA-2192            WA-2345  49361938   \n",
       "\n",
       "        latitude  longitude  latitude2  longitude2  \n",
       "0          6.304     -1.682      6.058      -1.391  \n",
       "1          5.970     -1.697      6.016      -1.813  \n",
       "2          6.304     -1.682      5.970      -1.697  \n",
       "3          5.971     -1.779      5.971      -1.778  \n",
       "4          5.970     -1.697      6.016      -1.812  \n",
       "...          ...        ...        ...         ...  \n",
       "715687     5.970     -1.697      6.015      -1.813  \n",
       "715688     5.991     -1.370      5.970      -1.779  \n",
       "715689     6.049     -1.961      6.047      -1.961  \n",
       "715690     6.369     -1.464      6.108      -1.449  \n",
       "715691     6.404     -1.451      6.207      -1.692  \n",
       "\n",
       "[715692 rows x 10 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dblton\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = allel.GenotypeArray(snps)\n",
    "contig = '2L'\n",
    "dblton = dblton.query(\"contig == @contig\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "index 0 is out of bounds for axis 0 with size 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_5291/1503534520.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'timeit'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'f2hapSize = {}\\n\\nfor idx, row in dblton[250_000:250_100].iterrows():\\n\\n    # take one doubleton at a time, get contig, pos, individual1 and 2\\n    contig, dbltonpos, idx1, idx2 = row[[\\'contig\\', \\'pos\\', \\'idx1\\', \\'idx2\\']]\\n    ## subset to individuals, we need whole chrom atm\\n\\n    # subset genotypes to these individuals\\n    geno1 = snps[:, idx1]\\n    geno2 = snps[:, idx2]\\n    # get boolean of dblton idx\\n    dblton_idx = np.where(pos == dbltonpos)[0][0]\\n\\n    lowerBreakpoint = dblton_idx.copy()\\n    upperBreakpoint = dblton_idx.copy()\\n\\n    # Reset position for next loop\\n    gn1 = geno1[dblton_idx]\\n    gn2 = geno2[dblton_idx]\\n    # Scan right along genome, as long as two inds are not both homozygous but different\\n    while not (gn1[0] == gn1[1]) & (gn2[0] == gn2[1]) & ((gn1 != gn2).all()) & (-1 not in gn1 and -1 not in gn2):\\n\\n        upperBreakpoint += 1\\n        if upperBreakpoint == max: # limit the upper breakpoint at end of the contig\\n            break\\n\\n        gn1 = geno1[upperBreakpoint]\\n        gn2 = geno2[upperBreakpoint]\\n\\n    # Scan left along genome\\n    # subset genotypes to this position, because we need somethign to start the while loop?\\n    gn1 = geno1[dblton_idx]\\n    gn2 = geno2[dblton_idx]\\n\\n    # Scan left along genome\\n    while not (gn1[0] == gn1[1]) & (gn2[0] == gn2[1]) & ((gn1 != gn2).all()) & (-1 not in gn1 and -1 not in gn2):\\n\\n        lowerBreakpoint -= 1\\n        if lowerBreakpoint == 0: # limit lower breakpoint at zero, start of contig\\n            break\\n\\n        gn1 = geno1[lowerBreakpoint]\\n        gn2 = geno2[lowerBreakpoint]\\n    \\n\\n    start = pos[lowerBreakpoint]\\n    end = pos[upperBreakpoint]\\n    f2hapSize[f\"{contig}_{dbltonpos}\"] = end-start  # get size in bp of f2 haplotype\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2417\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2418\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2419\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2420\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2421\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/decorator.py\u001b[0m in \u001b[0;36mfun\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    230\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mkwsyntax\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m                 \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 232\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mcaller\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextras\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    233\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m   1178\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1179\u001b[0m                 \u001b[0mnumber\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m \u001b[0;34m**\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1180\u001b[0;31m                 \u001b[0mtime_number\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnumber\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1181\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtime_number\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m0.2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1182\u001b[0m                     \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/lib/python3.7/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, number)\u001b[0m\n\u001b[1;32m    167\u001b[0m         \u001b[0mgc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 169\u001b[0;31m             \u001b[0mtiming\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    170\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgcold\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<magic-timeit>\u001b[0m in \u001b[0;36minner\u001b[0;34m(_it, _timer)\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: index 0 is out of bounds for axis 0 with size 0"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "f2hapSize = {}\n",
    "\n",
    "for idx, row in dblton[250_000:250_100].iterrows():\n",
    "\n",
    "    # take one doubleton at a time, get contig, pos, individual1 and 2\n",
    "    contig, dbltonpos, idx1, idx2 = row[['contig', 'pos', 'idx1', 'idx2']]\n",
    "    ## subset to individuals, we need whole chrom atm\n",
    "\n",
    "    # subset genotypes to these individuals\n",
    "    geno1 = snps[:, idx1]\n",
    "    geno2 = snps[:, idx2]\n",
    "    # get boolean of dblton idx\n",
    "    dblton_idx = np.where(pos == dbltonpos)[0][0]\n",
    "\n",
    "    lowerBreakpoint = dblton_idx.copy()\n",
    "    upperBreakpoint = dblton_idx.copy()\n",
    "\n",
    "    # Reset position for next loop\n",
    "    gn1 = geno1[dblton_idx]\n",
    "    gn2 = geno2[dblton_idx]\n",
    "    # Scan right along genome, as long as two inds are not both homozygous but different\n",
    "    while not (gn1[0] == gn1[1]) & (gn2[0] == gn2[1]) & ((gn1 != gn2).all()) & (-1 not in gn1 and -1 not in gn2):\n",
    "\n",
    "        upperBreakpoint += 1\n",
    "        if upperBreakpoint == max: # limit the upper breakpoint at end of the contig\n",
    "            break\n",
    "\n",
    "        gn1 = geno1[upperBreakpoint]\n",
    "        gn2 = geno2[upperBreakpoint]\n",
    "\n",
    "    # Scan left along genome\n",
    "    # subset genotypes to this position, because we need somethign to start the while loop?\n",
    "    gn1 = geno1[dblton_idx]\n",
    "    gn2 = geno2[dblton_idx]\n",
    "\n",
    "    # Scan left along genome\n",
    "    while not (gn1[0] == gn1[1]) & (gn2[0] == gn2[1]) & ((gn1 != gn2).all()) & (-1 not in gn1 and -1 not in gn2):\n",
    "\n",
    "        lowerBreakpoint -= 1\n",
    "        if lowerBreakpoint == 0: # limit lower breakpoint at zero, start of contig\n",
    "            break\n",
    "\n",
    "        gn1 = geno1[lowerBreakpoint]\n",
    "        gn2 = geno2[lowerBreakpoint]\n",
    "    \n",
    "\n",
    "    start = pos[lowerBreakpoint]\n",
    "    end = pos[upperBreakpoint]\n",
    "    f2hapSize[f\"{contig}_{dbltonpos}\"] = end-start  # get size in bp of f2 haplotype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hap Length Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Where do we find outliers? Lets plot f2 haplotype size against f2 haplotype (start) position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "f2df = pd.concat(f2Haps, axis=0).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>dblton_pos</th>\n",
       "      <th>contig</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>4689</td>\n",
       "      <td>81</td>\n",
       "      <td>2L</td>\n",
       "      <td>4673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>4652</td>\n",
       "      <td>82</td>\n",
       "      <td>2L</td>\n",
       "      <td>4636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>9158</td>\n",
       "      <td>123</td>\n",
       "      <td>2L</td>\n",
       "      <td>9142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16</td>\n",
       "      <td>12351</td>\n",
       "      <td>142</td>\n",
       "      <td>2L</td>\n",
       "      <td>12335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>399</td>\n",
       "      <td>145</td>\n",
       "      <td>2L</td>\n",
       "      <td>383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3836743</th>\n",
       "      <td>24393019</td>\n",
       "      <td>24393056</td>\n",
       "      <td>24393039</td>\n",
       "      <td>X</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3836744</th>\n",
       "      <td>24392903</td>\n",
       "      <td>24393106</td>\n",
       "      <td>24393044</td>\n",
       "      <td>X</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3836745</th>\n",
       "      <td>24370076</td>\n",
       "      <td>24393107</td>\n",
       "      <td>24393048</td>\n",
       "      <td>X</td>\n",
       "      <td>23031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3836746</th>\n",
       "      <td>24393056</td>\n",
       "      <td>24393107</td>\n",
       "      <td>24393059</td>\n",
       "      <td>X</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3836747</th>\n",
       "      <td>24392903</td>\n",
       "      <td>24393107</td>\n",
       "      <td>24393069</td>\n",
       "      <td>X</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3836748 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            start       end  dblton_pos contig   size\n",
       "0              16      4689          81     2L   4673\n",
       "1              16      4652          82     2L   4636\n",
       "2              16      9158         123     2L   9142\n",
       "3              16     12351         142     2L  12335\n",
       "4              16       399         145     2L    383\n",
       "...           ...       ...         ...    ...    ...\n",
       "3836743  24393019  24393056    24393039      X     37\n",
       "3836744  24392903  24393106    24393044      X    203\n",
       "3836745  24370076  24393107    24393048      X  23031\n",
       "3836746  24393056  24393107    24393059      X     51\n",
       "3836747  24392903  24393107    24393069      X    204\n",
       "\n",
       "[3836748 rows x 5 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f2df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "f2haps['dist_bin'] = pd.cut(f2haps['distance'], 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "f2haps['dist_bin_factor'] = pd.factorize(f2haps['dist_bin'])[0]"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "710100144729ec1057471750b60f3250df4296be7eb29e7e87fec32ed718ec4b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
